system/temperature.c

```diff
@@ -21,6 +21,141 @@
 
 #include "temperature.h"
 
+static uint16_t pci_read_pid(int bus, int dev, int func)
+{
+    return pci_config_read16(bus, dev, func, 2);
+}
+
+static uint16_t pci_read_vid(int bus, int dev, int func)
+{
+    return pci_config_read16(bus, dev, func, 0);
+}
+
+static int k8dev, k8fun;
+
+// find the first PCI device with the related IDs
+// On a multi-socketed K8, it grab only the first one since memtest only show
+// one temp
+static bool find_k8_temp(void)
+{
+    uint32_t pid, vid;
+
+    for (k8dev = 0; k8dev < 32; k8dev++) {
+        for (k8fun = 0; k8fun < 8; k8fun++) {
+            vid = pci_read_vid(0, k8dev, k8fun);
+            pid = pci_read_pid(0, k8dev, k8fun);
+            if (vid == PCI_VID_AMD && pid == 0x1103)
+                return true;
+        }
+    }
+    return false;
+}
+
+static int k8_temp(void)
+{
+    // k8temp is stored on a PCI device ([AMD] K8 [Athlon64/Opteron] Miscellaneous Control)
+#define K8_REG_TEMP   0xe4
+#define K8_SEL_CORE   0x04
+    uint32_t reg;
+    int temp = 0;
+
+    if (!find_k8_temp())
```

Comment:
Besides the potential merge of `find_k8_temp()` and `find_k10_temp()` I mentioned in another comment, this implementation of `k8_temp()` + `find_k8_temp()` and `k10_temp()` + `find_k10_temp()` reads the PCI bus config space every time, without caching the result of the enumeration. I can live with spending several bytes to avoid this repeated enumeration :)

Reply:
it is why k8dev,k8fun are static, I just forgot to add a if (!=0)return